### ISSUE TITLE
Large size of plots saved as pdf

### ISSUE THREAD

**Issue by fernando-goncalves** [matplotlib/matplotlib#9716]
### Bug report

**Bug summary**

matplotlib version 2.1.0 generates pdf files that are a lot larger than those generated by version 2.0.2. In one case a file that looks exactly the same went from 635KB to 127MB. A simpler code sample is presented below in which the size grew from 11KB to 20KB.
Is there a way to get the size of version 2.0.2?

**Code for reproduction**

```python
import matplotlib as mlp
import matplotlib.pyplot as plt

from matplotlib.backends.backend_pdf import PdfPages
from sklearn.datasets import load_iris

print(mlp.__version__)
iris = load_iris()
data = iris.data

plt.scatter(data[:,0], data[:,1])
plt.xlabel('x values')
plt.ylabel('y values')
plt.title('example')

pdf = PdfPages(r"test.pdf")
plt.savefig(pdf, format='pdf', bbox_inches = 'tight')
plt.show()
pdf.close()
pdf = None
```

**Matplotlib version**

  * Operating system:
  * Matplotlib version: 2.1.0
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version:
  * Jupyter version (if applicable):
  * Other libraries: 

conda

anaconda

**Edit (@afvincent):** cosmetic edit, mainly to get syntax highlighting.

**Comment by afvincent** [comment]
Thank you for the report :).

Here is a MWE that is a bit more self-contained (e.g. not requiring scikit-learn):
```python
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from matplotlib.backends.backend_pdf import PdfPages

version = mpl.__version__

prng = np.random.RandomState(19680107)
data = prng.random_sample(size=(150, 2))

fig, ax = plt.subplots()
ax.scatter(data[:,0], data[:,1])

pdf = PdfPages("test_{0}.pdf".format(version))
plt.savefig(pdf, format='pdf', bbox_inches='tight')
pdf.close()
pdf = None
```
It seems to generate a PDF that weights either ~8.6 kB (2.0.2) or ~13.8 kB (2.1.0).

A `git bisect` session with the previous MWE seems to designate the “culprit” commit (i.e. significantly increasing the weight of the PDF) to be:
```git
433b8996347dad2c144000a1d1e7b6b1e50f7cbe is the first bad commit
commit 433b8996347dad2c144000a1d1e7b6b1e50f7cbe
Author: Jouni K. Seppänen <jks@iki.fi>
Date:   Sun May 15 18:50:49 2016 +0300

    Separate pdf finalization from closing
    
    If an error is raised and the object state is inconsistent,
    we shouldn't try to finalize the file, just close all resources.
    The idea is to support the following idiom:
    
        try:
            figure.draw(file)  # write pdf file, can raise
            file.finalize()    # do this if everything went well
        finally:
            file.close()       # do this in any case
```
Pinging @jkseppan based on these results ;).

**Comment by afvincent** [comment]
For the moment, labelling it as release critical: producing PDF files of 100+ MB instead of less than 1 MB as in 2.0.2 seems to be a regression with potentially a very significant impact on end-users. *As always, if a core dev disagrees, feel free to remove the label or re-milestone the ticket :).*

**Comment by tacaswell** [comment]
@fernando-goncalves What is in the plot that got bigger by 20x?  Plots that get bigger by a factor of 2 are not really of the same concern.

Do the other vector backends share the same problem?

**Comment by jkseppan** [comment]
I'm a little low on bandwidth this week, but please feel free to remind me later. I agree with @tacaswell that it would be interesting to see an example where the size difference is much greater than 2x, because otherwise we might end up chasing a lower-order bug.

One possible workaround is to process the pdf files with Ghostscript. The command line options may be a bit nonobvious, but there are some good answers at https://stackoverflow.com/q/10450120/26575

**Comment by fernando-goncalves** [comment]
I noticed the issue when running a function that produces line plots for multiple variables and saves them to a single file. Each plot contains a line for each class in the data. In the case in point, there are only two classes, so each chart has two lines, and no chart has more than 50 data points (some have very few).

The following code reproduces the issue:
```
import os

import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from matplotlib.backends.backend_pdf import PdfPages

version = mpl.__version__

np.random.seed(0)
x = range(50)

size = {}
for f in [1, 10, 50, 100, 200, 300, 400]:
    pdf = PdfPages(r"savefig\test_{} size {}.pdf".format(version, f))
    for i in range (f):
        y1 = np.random.randint(0,100,size=50, )
        y2 = np.random.randint(0,100,size=50, )
        plt.plot(x, y1, label="label_y1 (100%, 30%)")
        plt.plot(x, y2, label="label_y2 (90%, 80%)")
        plt.legend(loc='best')
        plt.title("chart with long title for testing: {}".format(i))
        plt.savefig(pdf, format='pdf', bbox_inches='tight')
        plt.show()
    pdf.close()
    pdf = None
    size[f] = os.path.getsize(r"savefig\test_{} size {}.pdf".format(version, f))/1024

for k, v in size.items():
    print ("#charts: {:3} size(KB)={:8,.0f}  KB per chart={:5,.0f})".format(k, v, v/k))
```

it produces the following output (charts omitted):

```
#charts:   1 size(KB)=      27  KB per chart=   27)
#charts:  10 size(KB)=     195  KB per chart=   20)
#charts:  50 size(KB)=   1,743  KB per chart=   35)
#charts: 100 size(KB)=   5,499  KB per chart=   55)
#charts: 200 size(KB)=  19,085  KB per chart=   95)
#charts: 300 size(KB)=  40,772  KB per chart=  136)
#charts: 400 size(KB)=  70,567  KB per chart=  176)
```

Th problem appears to be cumulative and worsens the more text there is in the title and labels. When i increased the title to:

```
        plt.title("chart with even longer title for testing size impact: {}".format(i))
```
the output increased to:
```
#charts:   1 size(KB)=      29  KB per chart=   29)
#charts:  10 size(KB)=     211  KB per chart=   21)
#charts:  50 size(KB)=   1,896  KB per chart=   38)
#charts: 100 size(KB)=   6,001  KB per chart=   60)
#charts: 200 size(KB)=  20,870  KB per chart=  104)
#charts: 300 size(KB)=  44,624  KB per chart=  149)
#charts: 400 size(KB)=  77,267  KB per chart=  193)
```

As mentioned earlier, with the actual data and function I'm using, the size grew from 635KB to 127MB, for versions 2.0.2 and 2.1.0 respectively.

**Comment by jenshnielsen** [comment]
@fernando-goncalves did you intentionally close this?

**Comment by fernando-goncalves** [comment]
No. I must have done it accidentally.

**Comment by jklymak** [comment]
So this is a multi-page PDF issue, not a single-page PDF issue.

**Comment by fernando-goncalves** [comment]
Indeed, it is a multi-page pdf. I should have been more specific. I followed jkseppan's suggestion and used Ghostscript to reduce the 127 MB file to 2.3 MB, which is not as good as the original 635 KB, but is good enough for my purposes.

I agree that this is not critical and hopefully can be addressed in a future release.

**Comment by jklymak** [comment]
Well, I guess we can leave as 2.1.1 for now and see if it gets fixed.  There is clearly something funky going on.  But I'll remove the Release Critical so the core devs can re-milestone on if they deem appropriate.

**Comment by tacaswell** [comment]
```python
import os

import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from matplotlib.backends.backend_pdf import PdfPages

version = mpl.__version__

np.random.seed(0)
x = range(50)

size = {}
for f in [1, 10, 50, 100, 200, 300, 400]:
    fname = r"savefig_test_{} size {}_2.pdf".format(version, f)
    fig, ax = plt.subplots()
    pdf = PdfPages(fname)
    for i in range(f):
        ax.cla()
        y1 = np.random.randint(0, 100, size=50)
        y2 = np.random.randint(0, 100, size=50)
        ax.plot(x, y1, label="label_y1 (100%, 30%)")
        ax.plot(x, y2, label="label_y2 (90%, 80%)")
        ax.legend(loc='best')
        ax.set_title("chart with long title for testing: {}".format(i))
        fig.savefig(pdf, format='pdf')

    pdf.close()
    pdf = None
    size[f] = os.path.getsize(fname)/1024

for k, v in size.items():
    print("#charts: {:3} size(KB)={:8,.0f}  KB per chart={:5,.0f})"
          .format(k, v, v/k))

```

2.0.2

```
12:33 $ python mpl_scratch.py 
#charts:   1 size(KB)=      14  KB per chart=   14)
#charts:  10 size(KB)=      32  KB per chart=    3)
#charts:  50 size(KB)=     111  KB per chart=    2)
#charts: 100 size(KB)=     210  KB per chart=    2)
#charts: 200 size(KB)=     409  KB per chart=    2)
#charts: 300 size(KB)=     607  KB per chart=    2)
#charts: 400 size(KB)=     804  KB per chart=    2)

```

2.1.x(ish)

```
(dd36) ✔ /tmp 
12:37 $ python mpl_scratch.py 
#charts:   1 size(KB)=      27  KB per chart=   27)
#charts:  10 size(KB)=     195  KB per chart=   19)
#charts:  50 size(KB)=   1,741  KB per chart=   35)
#charts: 100 size(KB)=   5,495  KB per chart=   55)
#charts: 200 size(KB)=  19,078  KB per chart=   95)
#charts: 300 size(KB)=  40,761  KB per chart=  136)
#charts: 400 size(KB)=  70,551  KB per chart=  176)

```

**Comment by tacaswell** [comment]
This is a pretty major regression.  We won't due 2.1.1 before this weekend (I am physically in a different state than my gpg key) so we have a bit of time to look into this.

**Comment by tacaswell** [comment]
We clearly have a line-painter-problem type bug.  In the pdf file from 2.0.2 there is one line that looks like

```
<< /Type /Pages
/Kids [ 10 0 R 13 0 R 16 0 R 19 0 R 22 0 R 25 0 R 28 0 R 31 0 R 34 0 R 37 0 R ]
/Count 10 >>
```

where as in the files from 2.1 have N +1, ex

```
(dd36) ✔ /tmp 
13:00 $ grep -a -C 1 Kids *10_2.pdf
savefig_test_2.0.2 size 10_2.pdf-<< /Type /Pages
savefig_test_2.0.2 size 10_2.pdf:/Kids [ 10 0 R 13 0 R 16 0 R 19 0 R 22 0 R 25 0 R 28 0 R 31 0 R 34 0 R 37 0 R ]
savefig_test_2.0.2 size 10_2.pdf-/Count 10 >>
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 1 /Kids [ 10 0 R ] /Type /Pages >>
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-endobj
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 2 /Kids [ 10 0 R 50 0 R ] /Type /Pages >>
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-endobj
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 3 /Kids [ 10 0 R 50 0 R 90 0 R ] /Type /Pages >>
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-endobj
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 4 /Kids [ 10 0 R 50 0 R 90 0 R 130 0 R ] /Type /Pages >>
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-endobj
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 5 /Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R ] /Type /Pages >>
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-endobj
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 6 /Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R 210 0 R ]
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-/Type /Pages >>
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-2 0 obj
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:<< /Count 7 /Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R 210 0 R 250 0 R ]
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-/Type /Pages >>
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-<< /Count 8
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:/Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R 210 0 R 250 0 R 290 0 R ]
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-/Type /Pages >>
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-<< /Count 9
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:/Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R 210 0 R 250 0 R 290 0 R 331 0 R ]
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-/Type /Pages >>
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-<< /Count 10
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:/Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R 210 0 R 250 0 R 290 0 R 331 0 R
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-372 0 R ]
--
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-<< /Count 10
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf:/Kids [ 10 0 R 50 0 R 90 0 R 130 0 R 170 0 R 210 0 R 250 0 R 290 0 R 331 0 R
savefig_test_2.1.0.post210+gaf904fc3d size 10_2.pdf-372 0 R ]

```

**Comment by jklymak** [comment]
@fernando-goncalves I think #9724 fixes the problem.  If you have the ability to run a development branch it'd be great if you gave it a try (but I understand if you can't).
